datasets/voc/voc.py

```diff
@@ -0,0 +1,229 @@
+        height = float(size.find("height").text)
+
+        for obj in root.findall("object"):
+            # Get object's label name.
+            label = obj.find("name").text.lower()
+            # Get objects' pose name.
+            pose = obj.find("pose").text.lower()
+            is_truncated = obj.find("truncated").text == "1"
+            is_difficult = obj.find("difficult").text == "1"
+            bndbox = obj.find("bndbox")
+            xmax = float(bndbox.find("xmax").text)
+            xmin = float(bndbox.find("xmin").text)
+            ymax = float(bndbox.find("ymax").text)
+            ymin = float(bndbox.find("ymin").text)
+            yield {
+                "label": label,
+                "pose": pose,
+                "bbox": [ymin / height, xmin / width, ymax / height, xmax / width],
+                "is_truncated": is_truncated,
+                "is_difficult": is_difficult,
+            }
+
+
+class Voc(datasets.GeneratorBasedBuilder):
+    """Pascal VOC dataset."""
+
+    BUILDER_CONFIGS = [
+        VocConfig(
+            year="2007",
+            description=_VOC_CONFIG_DESCRIPTION.format(year=2007, num_images=9963, num_objects=24640),
+            filenames={
+                "trainval": "VOCtrainval_06-Nov-2007.tar",
+                "test": "VOCtest_06-Nov-2007.tar",
+            },
+            has_test_annotations=True,
+        ),
+        VocConfig(
+            year="2012",
+            description=_VOC_CONFIG_DESCRIPTION.format(year=2012, num_images=11540, num_objects=27450),
+            filenames={
+                "trainval": "VOCtrainval_11-May-2012.tar",
+                "test": "VOC2012test.tar",
+            },
+            has_test_annotations=False,
+        ),
+    ]
+
+    # DEFAULT_CONFIG_NAME = "2012"
+
+    def _info(self):
+        return datasets.DatasetInfo(
+            description=_VOC_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "image": datasets.Image(),
+                    "objects": datasets.Sequence(
+                        {
+                            "label": datasets.ClassLabel(names=_VOC_LABELS),
+                            "bbox": datasets.Sequence(datasets.Value("float32"), length=4),
+                            "pose": datasets.ClassLabel(names=_VOC_POSES),
+                            "is_truncated": datasets.Value("bool"),
+                            "is_difficult": datasets.Value("bool"),
+                        }
+                    ),
+                    "labels": datasets.Sequence(datasets.ClassLabel(names=_VOC_LABELS)),
+                    "labels_no_difficult": datasets.Sequence(datasets.ClassLabel(names=_VOC_LABELS)),
+                }
+            ),
+            homepage=_VOC_URL.format(year=self.config.year),
+            citation=_VOC_CITATION.format(year=self.config.year),
+        )
+
+    def _split_generators(self, dl_manager):
+        paths = dl_manager.download_and_extract(
```

Comment:
To be able to make this dataset work in streaming mode, you have to first `dl_manager.download` the TAR archive, and then iterate over the files using `dl_manager.iter_archive`.

Unfortunately it's not possible to `dl_manager.download_and_extract` TAR archives in streaming mode.

Streaming mode is also useful for the dataset preview on the Hub.

If you need certain files to be hosted on the Hub let me know and I can help you with it :)

Reply:
Can the txt files fit in memory ? If so you can just load them all in a python dictionary for the lookup of your first proposition. I think 

(otherwise regarding option 2: the images come from Flickr (see terms of use [here](https://www.flickr.com/help/terms)), flickr users retain their copyright on the content of flickr unless they specify an open license like Creative Commons - but this is not the general case. As for ImageNet I think we could redistribute it only if we make this dataset a gated dataset)

Reply:
I also think we should only go with the 2nd option + gated access if the txt files don't fit in memory.

Reply:
The txt files would fit in memory, but it still doesn't solve the problem that the files I need to reference are in multiple directories, so iterating over the whole archive doesn't make sense I think. You can see here that I'd use the names in `train.txt` or `val.txt` to reference files in the `Annotations` and `JPEGImages` directories.

<img width="293" alt="Screen Shot 2022-06-13 at 1 54 45 PM" src="https://user-images.githubusercontent.com/32437151/173414882-0d540c08-2f7c-44f3-b66e-66f7e2da224f.png">